File: lr_test/first_follow1.js
// https://github.com/MikeDevice/first-follow

// 参考: https://bbs.pediy.com/thread-248501.htm

const EMPTY_CHAIN = null;
const END_MARKER = '$';

function firstFollow(rules) {
    let firstSets = {};
    let followSets = {};
    let predictSets = {};

    rules.map(({ left }) => {
        firstSets[left] = [];
        followSets[left] = [];
    });

    function union(arr1, arr2) {
        return [...new Set([...arr1, ...arr2])];
    }

    function isNonterminal(item) {
        return firstSets[item];
    }

    function collectSet(initialSet, items, additionalSet) {
        let set = initialSet;
        // NOTE every和forEach语义差异很大: every是做每项检查，默认返回为false，即不明确返回true即中止
        items.every((item, index) => {
            if (isNonterminal(item)) {
                set = union(set, firstSets[item].filter((setItem) => setItem !== EMPTY_CHAIN));

                if (firstSets[item].includes(EMPTY_CHAIN)) {
                    if (items[index + 1]) return true;
                    set = union(set, additionalSet);
                }
            } else {
                set = union(set, [item]);
                return false
            }
        });

        return set;
    }

    function makeFirstSets() {
        let isSetChanged;

        do {
            isSetChanged = false;

            rules.forEach(({ left, right }) => {
                let set = firstSets[left];
                let collect_set = collectSet(set, right, [EMPTY_CHAIN])
                set = union(set,  collect_set);

                if (firstSets[left].length !== set.length) {
                    firstSets[left] = set;
                    isSetChanged = true;
                }
            });
        } while (isSetChanged);

        return firstSets;
    }

    // 唯一搞定那个怪语法且有源码的! -- 这个看着像虎书算法，不是龙书那个混淆定义！
    function makeFollowSets() {
        followSets[rules[0].left].push(END_MARKER);

        let isSetChanged;

        do {
            isSetChanged = false;

            rules.forEach(({ left, right }) => {
                right.forEach((item, index) => {
                    if (!isNonterminal(item)) return;

                    let set = followSets[item];

                    // NOTE 就这么一句就算好了
                    set = union(
                        set,
                        index + 1 < right.length
                            ? collectSet(set, right.slice(index + 1), followSets[left])
                            : followSets[left],
                    );

                    if (followSets[item].length !== set.length) {
                        followSets[item] = set;
                        isSetChanged = true;
                    }
                });
            });
        } while (isSetChanged);

        return followSets;
    }

    function makePredictSets() {
        rules.forEach(({ left, right }, ruleIndex) => {
            const firstItem = right[0];
            let set = [];

            if (isNonterminal(firstItem)) {
                set = union(set, collectSet(set, right, followSets[left]));
            } else if (firstItem === EMPTY_CHAIN) {
                set = [...followSets[left]];
            } else {
                set.push(firstItem);
            }

            predictSets[ruleIndex + 1] = set;
        });

        return predictSets;
    }

    firstSets = makeFirstSets();
    followSets = makeFollowSets();
    predictSets = makePredictSets();

    return { firstSets, followSets, predictSets };
};


let rules = [
    {
        left: 'S',
        right: ['A', 'C', 'B']
    },
    {
        left: 'S',
        right: ['C', 'b', 'b']
    },
    {
        left: 'S',
        right: ['B', 'a']
    },
    {
        left: 'A',
        right: ['d', 'a']
    },
    {
        left: 'A',
        right: ['B', 'C']
    },
    {
        left: 'B',
        right: ['g']
    },
    {
        left: 'B',
        right: [null]
    },
    {
        left: 'C',
        right: ['h']
    },
    {
        left: 'C',
        right: [null]
    },
];

// 问题, 如果用字面'ϵ'而不用null, follow集计算结果不正确
rules = [
    {left:"E", right:["T","E'"]},
    {left:"E'",right:["+","T","E'"]},
    {left:"E'",right:[null]},
    {left:"T", right:["F","T'"]},
    {left:"T'",right:["*","F","T'"]},
    {left:"T'",right:[null]},
    {left:"F", right:["id"]},
    {left:"F", right:["(","E",")"]},       
];

const { firstSets, followSets, predictSets } = firstFollow(rules);

// console.log(rules)
console.log(firstSets)
console.log(followSets)
console.log(predictSets)

File: lr_test/first_follow1.py
# https://github.com/MikeDevice/first-follow

from typing import List
from pprint import pprint

rules = [
  { 'left': 'S', "right": [ 'A', 'C', 'B' ] },
  { 'left': 'S', "right": [ 'C', 'b', 'b' ] },
  { 'left': 'S', "right": [ 'B', 'a' ] },
  { 'left': 'A', "right": [ 'd', 'a' ] },
  { 'left': 'A', "right": [ 'B', 'C' ] },
  { 'left': 'B', "right": [ 'g' ] },
  { 'left': 'B', "right": [ 'ϵ' ] },
  { 'left': 'C', "right": [ 'h' ] },
  { 'left': 'C', "right": [ 'ϵ' ] }
]

firstSets = {}
followSets = {}
isNoTerm = lambda item:item.isupper()

def getSet(dict,key)->set:
    if key not in dict:
        dict[key] = set()
    return dict[key]


# set_:         旧的目标fristSet|followSet
# items:        右侧项集(follow会算递减子集)
# additionSet:  两种情况不同, firt->[ϵ], follow->右侧完整项集
def collectSet(set_:set,items:List[str],additionSet:List[str])->set:
    # 下面的continue和break模仿的是js的every语义
    for idx,item in enumerate(items):
        # 如非终接符 循环处理
        if isNoTerm(item):
            item_first_set = getSet(firstSets,item)
            set_ = set_.union( [sym for sym in item_first_set if sym != 'ϵ'] ) 
            # 如果这个ϵ能一直串接下去
            if 'ϵ' in item_first_set:
                # 后面还有字符继续
                if len(items) > idx + 1 and items[idx + 1]: 
                    continue
                # 后面没有字符，对first集就加上ϵ
                set_ = set_.union( additionSet )
            # 这里不能省, 出现过bug !!!
            else:
                break                   
        # 如终接符 只取一个
        else:
            set_ = set_.union( [item] ) 
            break
    return set_

def makeFirstSets():
    isSetChanged:bool = True
    while isSetChanged:
        isSetChanged = False
        for rule in rules:
            left,right = rule['left'],rule['right']
            firstSet = getSet(firstSets,left)
            firstSet = collectSet( firstSet, right, ['ϵ'] )
            if len( firstSets[left] ) != len( firstSet ):
                firstSets[left] = firstSet
                isSetChanged = True
    return firstSets

def makeFollowSets():
    getSet(followSets,rules[0]['left']).add('$') # 为起始项添加 '$'

    isSetChanged:bool = True
    while isSetChanged:
        isSetChanged = False
        for rule in rules:
            left,right = rule['left'],rule['right']
            for idx,item in enumerate(right):
                if not isNoTerm(item): continue
                followSet = getSet(followSets,item)

                if idx + 1 < len(right):
                    temp = collectSet( followSet , right[idx+1:], getSet(followSets,left) ) #注意slice语义
                else:
                    temp = getSet(followSets,left)
                followSet = followSet.union( temp )

                if len( followSets[item] ) != len( followSet ):
                    followSets[item] = followSet
                    isSetChanged = True
    return followSets                

makeFirstSets()
pprint(firstSets)
makeFollowSets()
pprint(followSets)